linkerd/app/outbound/src/protocol.rs

```diff
@@ -0,0 +1,142 @@
+            Error = Error,
+        >,
+        HSvc: Clone + Send + Sync + Unpin + 'static,
+        HSvc::Future: Send,
+        // Opaque connection stack.
+        N: svc::NewService<T, Service = NSvc>,
+        N: Clone + Send + Sync + Unpin + 'static,
+        NSvc: svc::Service<io::EitherIo<I, io::PrefixedIo<I>>, Response = (), Error = Error>,
+        NSvc: Clone + Send + Sync + Unpin + 'static,
+        NSvc::Future: Send,
+    {
+        let opaq = self.clone().into_stack();
+
+        let http = self.with_stack(http).map_stack(|config, rt, stk| {
+            stk.push_on_service(http::BoxRequest::layer())
+                .unlift_new()
+                .push(http::NewServeHttp::layer(
+                    config.proxy.server.h2_settings,
+                    rt.drain.clone(),
+                ))
+        });
+
+        // The inner stacks are built for each connection and not cached.
+        let detect = http.clone().map_stack(|config, _, http| {
+            http.push_switch(
+                |(result, parent): (detect::Result<http::Version>, T)| -> Result<_, Infallible> {
+                    Ok(match detect::allow_timeout(result) {
+                        Some(version) => svc::Either::A(Http { version, parent }),
+                        None => svc::Either::B(parent),
+                    })
+                },
+                opaq.clone().into_inner(),
+            )
+            // `DetectService` oneshots the inner service, so we add
+            // a loadshed to prevent leaking tasks if (for some
+            // unexpected reason) the inner service is not ready.
+            .push_on_service(svc::LoadShed::layer())
+            .push_on_service(svc::MapTargetLayer::new(io::EitherIo::Right))
+            .lift_new_with_target::<(detect::Result<http::Version>, T)>()
+            .push(detect::NewDetectService::layer(config.proxy.detect_http()))
+        });
+
+        http.map_stack(|_, _, http| {
+            http.push_switch(Ok::<_, Infallible>, opaq.clone().into_inner())
+                .push_on_service(svc::MapTargetLayer::new(io::EitherIo::Left))
+                .push_switch(
+                    |parent: T| -> Result<_, Infallible> {
+                        match parent.param() {
+                            Protocol::Http1 => Ok(svc::Either::A(svc::Either::A(Http {
+                                version: http::Version::Http1,
+                                parent,
+                            }))),
+                            Protocol::Http2 => Ok(svc::Either::A(svc::Either::A(Http {
+                                version: http::Version::H2,
+                                parent,
+                            }))),
+                            Protocol::Opaque => Ok(svc::Either::A(svc::Either::B(parent))),
+                            Protocol::Detect => Ok(svc::Either::B(parent)),
+                        }
+                    },
+                    detect.into_inner(),
+                )
```

Comment:
the nested switch here (and corresponding two levels of nested of `Either`s) feels a bit complex, it might be nice if there was a comment explaining what the inner and outer `Either`s control at the top of the match or something?